test/field/RungeKutta.test.cc

```diff
@@ -0,0 +1,147 @@
+    }
+
+  protected:
+    // Test parameters
+    FieldTestParams param;
+};
+
+//---------------------------------------------------------------------------//
+// TESTS
+//---------------------------------------------------------------------------//
+
+TEST_F(RungeKuttaTest, rk4_host)
+{
+    // Construct the Runge-Kutta stepper
+    MagField                         field({0, 0, param.field_value});
+    FieldEquation                    equation(field);
+    RungeKuttaStepper<FieldEquation> rk4(equation);
+
+    // Test parameters and the sub-step size
+    real_type hstep = 2.0 * constants::pi * param.radius / param.nsteps;
+
+    for (CELER_MAYBE_UNUSED int i : celeritas::range(param.nstates))
+    {
+        // Initial state and the epected state after revolutions
+        Array<real_type, 6> y;
+        y[0] = param.radius;
+        y[1] = 0.0;
+        y[2] = i * 1.0e-6;
+        y[3] = 0.0;
+        y[4] = param.momentum_y;
+        y[5] = param.momentum_z;
+
+        Array<real_type, 6> expected_y;
+        for (std::size_t i = 0; i != 6; ++i)
+            expected_y[i] = y[i];
+
+        // The rhs of the equation and a temporary array
+        Array<real_type, 6> dydx;
+        Array<real_type, 6> yout;
+
+        // Try the stepper by hstep for (num_revolutions * num_steps) times
+        real_type total_err2 = 0;
+        for (int nr : range(param.revolutions))
+        {
+            // Travel hstep for num_steps times in the field
+            expected_y[2] = param.delta_z * (nr + 1) + i * 1.0e-6;
+            for (CELER_MAYBE_UNUSED int j : celeritas::range(param.nsteps))
+            {
+                dydx           = equation(y);
+                yout           = rk4(hstep, y, dydx);
+                real_type err2 = rk4.error(hstep, y);
+                y              = yout;
+                total_err2 += err2;
+            }
+            // Check the state after each revolution and the total error
+            EXPECT_VEC_NEAR(expected_y, y, sqrt(total_err2));
+            EXPECT_LT(total_err2, param.epsilon);
+        }
+    }
+}
+
+#if CELERITAS_USE_CUDA
+//---------------------------------------------------------------------------//
+// DEVICE TESTS
+//---------------------------------------------------------------------------//
+
+class RungeKuttaDeviceTest : public RungeKuttaTest
+{
+};
+
+TEST_F(RungeKuttaDeviceTest, rk4_device)
+{
+    // Run kernel
+    auto output = rk4_test(param);
+
+    // Check stepper results
+    real_type zstep = param.delta_z * param.revolutions;
+    for (unsigned int i = 0; i < output.pos_x.size(); ++i)
+    {
+        real_type error = std::sqrt(output.error[i]);
+        EXPECT_LT(fabs(output.pos_x[i] - param.radius), error);
+        EXPECT_LT(fabs(output.pos_z[i] - (zstep + i * 1.0e-6)), error);
+        EXPECT_LT(fabs(output.mom_y[i] - param.momentum_y), error);
+        EXPECT_LT(fabs(output.mom_z[i] - param.momentum_z), error);
```

Comment:
The `EXPECT_SOFT_EQ` (default of a few ULP) and `EXPECT_SOFT_NEAR` (with customizable relative error tolerance) might be better for this since they can give very informative error messages if the test fails, as well as being easier to read:
```suggestion
        EXPECT_SOFT_NEAR(output.mom_z[i], param.momentum_z, error);
```

Reply:
Replaced.